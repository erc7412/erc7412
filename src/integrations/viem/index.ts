export { createErc7412PublicActions } from './actions-public'
export { createErc7412WalletActions } from './actions-wallet'
export { createErc7412SendCalls } from './actions-sendCalls'
