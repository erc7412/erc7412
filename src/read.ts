import { OracleAdapter, TransactionRequest } from './types'
import type { SimulateCallsReturnType } from 'viem/actions'
import * as viem from 'viem'
import { resolvePrependTransaction } from './txn'

export async function simulateWithOffchainData<T extends unknown[]>(
  provider: Parameters<typeof viem.custom>[0],
  adapters: OracleAdapter[],
  transactions: TransactionRequest<T>,
  from: viem.Address = viem.zeroAddress,
  maxIter = 5
): Promise<{ results: SimulateCallsReturnType<T[]>['results']; txns: TransactionRequest<any>[] }> {
  const client = viem.createPublicClient({
    transport: viem.custom(provider, { retryCount: 0 })
  })

  let prependedTxns: TransactionRequest<{ to: viem.Address; data: viem.Hex; value: bigint }[]> = []
  for (let i = 0; i < maxIter; i++) {
    const simulatedCalls = await client.simulateCalls<any>({
      account: from,
      calls: [...prependedTxns, ...transactions]
    })

    console.log('simulated calls result', simulatedCalls)

    let batchNewTxs: typeof prependedTxns = []
    for (const result of simulatedCalls.results) {
      if (result.status !== 'success') {
        const err = result.error
        const newPrependTxs = await resolvePrependTransaction(err, client, adapters)
        batchNewTxs = [...batchNewTxs, ...newPrependTxs]
      }
    }

    if (!batchNewTxs.length) {
      return {
        results: simulatedCalls.results.slice(-transactions.length).map((r: any) => r.result) as any,
        txns: [...prependedTxns, ...transactions]
      }
    }

    prependedTxns = [...batchNewTxs, ...prependedTxns]
  }

  throw new Error('erc7412 callback repeat limit exceeded')
}
