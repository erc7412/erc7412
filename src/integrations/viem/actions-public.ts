import * as viem from 'viem'

import { estimateContractGas as actionEstimateContractGas } from 'viem/actions'

import { createErc7412WalletActions } from './actions-wallet'

import { simulateWithOffchainData } from '../../read'
import type { OracleAdapter } from '../../types'

export function getAccount (account: `0x${string}` | viem.Account | undefined): `0x${string}` {
  if (account === undefined) {
    return viem.zeroAddress
  }
  if (typeof account === 'string') {
    return viem.getAddress(account)
  }
  return viem.getAddress(account.address)
}

/**
 * Extend your viem client with the object returned by this function to automatically apply erc7412
 * required offchain data to your read calls
 */
export function createErc7412PublicActions (adapters: OracleAdapter[]) {
  const actionsWallet = createErc7412WalletActions(adapters)
  return (client: viem.PublicClient) => {
    const actions = {
      call: async (args: viem.CallParameters): Promise<viem.CallReturnType> => {
        return {
          data: (
            await simulateWithOffchainData(client, adapters, [
              {
                // from: getAccount(args.account),
                to: args.to || viem.zeroAddress,
                data: args.data || '0x',
                value: args.value || BigInt(0)
              }
            ])
          ).results[0].data
        }
      },
      readContract: async (args: viem.ReadContractParameters): Promise<viem.ReadContractReturnType> => {
        return {
          data: viem.decodeFunctionResult({
            ...args,
            data: (
              await simulateWithOffchainData(client, adapters, [
                {
                  data: viem.encodeFunctionData(args),
                  to: args.address || viem.zeroAddress,
                  value: BigInt(0)
                }
              ])
            ).results[0].data
          })
        }
      },
      prepareTransactionRequest: async (args: viem.PrepareTransactionRequestParameters) => {
        // the wallet has all the facilities necessary to proceess a possible transaction request
        return await actionsWallet(client).prepareTransactionRequest(args)
      },
      estimateContractGas: async (args: viem.EstimateContractGasParameters) => {
        try {
          return await actionEstimateContractGas(client, args)
        } catch (err) {
          console.log('WARN: erc7412 not implemented for estimateContractGas')
          throw err
        }
      },
      // TODO: types
      simulateContract: async (args: viem.SimulateContractParameters): Promise<any> => {
        return await actionsWallet(client).simulateContract(args)
      },
      multicall: async (args: viem.MulticallParameters): Promise<viem.MulticallReturnType> => {
        if (args.contracts.length < 1) {
          throw new Error('must have at least one call for multicall')
        }

        const retvals = await simulateWithOffchainData(
          // todo: types have a problem with the fact it cannot verify that the array is at least 1 long
          client,
          adapters,
          args.contracts.map((c) => {
            return {
              // from: c.address,
              data: viem.encodeFunctionData(c),
              to: c.address,
              value: BigInt(0)
            }
          })
        )

        return retvals.results.map((r, i) => {
          return { result: viem.decodeFunctionResult({ ...args.contracts[i], data: r.data }), status: 'success' }
        })
      }
      // below functions are not included because they are not applicable
      // writeContract: async (args: viem.WriteContractParameters) => {},
    }

    return actions
  }
}
