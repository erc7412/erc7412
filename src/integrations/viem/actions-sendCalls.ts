import * as viem from 'viem';
import { sendCalls as actionSendCalls } from 'viem/actions';
import type { OracleAdapter, SendCallsWithOffchainDataParameters } from '../../types';
import { simulateWithOffchainData } from '../../read';
import { getAccount } from './actions-public';

export function createErc7412SendCalls(adapters: OracleAdapter[]) {
  return (client: viem.WalletClient) => ({
    async sendCalls(
      args: SendCallsWithOffchainDataParameters,
    ): Promise<viem.SendCallsReturnType> {
      const { includeOffchainData, ...rest } = args as any;

      if (!includeOffchainData) {
        // fall back to native behaviour
        return actionSendCalls(client, rest);
      }

      // 1. work out the prepend transactions
      const fromAddress = getAccount(rest.account);
      const { txns } = await simulateWithOffchainData(
        client,                  // provider
        adapters,                // our oracle adapters
        rest.calls,              // user-provided bundle
        fromAddress,             // sender
      );

      // 2. hand off to stock sendCalls
      return actionSendCalls(client, {
        ...rest,
        calls: txns,             // prepend + user calls
        forceAtomic: true,       // Enforce atomic execution
      });
    },
  });
} 