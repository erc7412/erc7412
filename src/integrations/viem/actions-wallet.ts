import * as viem from 'viem'

import { prepareTransactionRequest as actionPrepareTransactionRequest } from 'viem/actions'

import { resolvePrependTransaction } from '../../txn'
import { simulateWithOffchainData } from '../../read'
import { makeTrustedForwarderMulticall } from '../../write'

import ITrustedMulticallForwarder from '../../../out/ITrustedMulticallForwarder.sol/ITrustedMulticallForwarder.json'

import type { TransactionRequest } from '../..'
import type { OracleAdapter } from '../../types'
import { getAccount } from './actions-public'

/**
 * Extend your viem client with the object returned by this function to automatically apply erc7412
 * required offchain data to your read calls
 */
export function createErc7412WalletActions(adapters: OracleAdapter[]) {
  return (client: viem.PublicClient) => {
    const actions = {
      prepareTransactionRequest: async (
        args: viem.PrepareTransactionRequestParameters
      ): Promise<viem.PrepareTransactionRequestReturnType> => {
        return await actions.prepareMulticallTransactionRequest({ txns: [args] })
      },

      prepareMulticallTransactionRequest: async <T extends unknown[]>(args: {
        txns: viem.PrepareTransactionRequestParameters[]
      }): Promise<viem.PrepareTransactionRequestReturnType> => {
        let prependedTxns: TransactionRequest<{}[]> = []

        if (args.txns.length < 1) {
          throw new Error('must have at least 1 transaction in multicall')
        }

        const payloadTxns: TransactionRequest<{ data: viem.Hex; to: viem.Address; value: bigint }[]> = args.txns.map((t) => {
          const req = {
            data: t.data || '0x',
            to: t.to || viem.zeroAddress,
            value: t.value || BigInt(0)
          }

          return req
        })

        while (true) {
          const multicallTxn: viem.PrepareTransactionRequestParameters =
            payloadTxns.length > 1 || prependedTxns.length > 0
              ? {
                  // TODO: is this doable without any?
                  ...makeTrustedForwarderMulticall([...prependedTxns, ...payloadTxns] as any[]),
                  account: args.txns[0].account,
                  chain: args.txns[0].chain
                }
              : args.txns[0]
          try {
            return await actionPrepareTransactionRequest(client, multicallTxn)
          } catch (err) {
            prependedTxns = [...prependedTxns, ...(await resolvePrependTransaction(err, client, adapters))]
          }
        }
      },
      simulateContract: async (args: viem.SimulateContractParameters): Promise<viem.SimulateContractReturnType> => {
        const baseTxn = {
          from: (args.account as any)?.address ?? viem.zeroAddress,
          to: args.address,
          chain: args.chain,
          data: viem.encodeFunctionData(args),
          value: args.value
        }
        const preparedTxn = await actions.prepareTransactionRequest(baseTxn)
        const execResult = await simulateWithOffchainData(client, adapters, [baseTxn])
        const txnData = preparedTxn.data

        if (preparedTxn.to !== args.address) {
          if (txnData === undefined) {
            throw new Error('prepared txn should have needed data for multicall')
          }
          return {
            request: {
              ...preparedTxn,
              abi: ITrustedMulticallForwarder.abi,
              address: preparedTxn.to,
              functionName: 'aggregate3Value',
              args: viem.decodeFunctionData({ abi: ITrustedMulticallForwarder.abi, data: txnData })
            } as any,
            result: viem.decodeFunctionResult({
              ...args,
              data: execResult.results[0].data
            }) as never
          }
        }

        console.log('execution result', execResult)
        return {
          request: args,
          result: viem.decodeFunctionResult({
            ...args,
            data: execResult.results[0].data
          })
        } as any // TODO
      }

      // NOTE: cant override `sendRawTransaction` because the transaction is
      // already serialized and so its impossible to impart a new multicall
      // in front without breaking permissions
      // `sendTransaction` is not included because its behavior is not supposed to affect the transaction,
      // its just supposed to sign and send
    }

    return actions
  }
}
