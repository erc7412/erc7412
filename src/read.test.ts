import * as mod from './read'

import * as viem from 'viem'

import IERC7412 from '../out/IERC7412.sol/IERC7412.0.8.27.json'

import type { OracleAdapter } from './types'

export const fakeWeb3 = {
  request: jest.fn()
}

export const fakeAdapters: OracleAdapter[] = [
  {
    getOracleId: () => 'FAKE',
    fetchOffchainData: async (_client, _oracleContract, oracleQuery: any[]) =>
      oracleQuery.map((v) => ({ arg: ('0x8765' + v.query.slice(2)) as viem.Hex, fee: BigInt(100) }))
  }
]

describe('read.ts', () => {
  let errorCode: viem.Hex = '0x1234'
  beforeEach(() => {
    errorCode = '0x1234'
    fakeWeb3.request.mockImplementation(async ({ method, params }: { method: string; params: any[] }) => {
      if (method === 'eth_chainId') {
        return BigInt(1337)
      } else if (method === 'eth_simulateV1') {
        console.log('block state calls', params[0].blockStateCalls[0])
        if (params[0].blockStateCalls[0].calls[0].data.includes('87651234') === true) {
          return [
            {
              baseFeePerGas: '0x0',
              blobGasUsed: '0x0',
              calls: [
                {
                  returnData: viem.encodeAbiParameters(
                    [viem.parseAbiParameter('(bool success, bytes returnData)[]')],
                    [
                      [
                        { success: true, returnData: '0x1234' },
                        { success: true, returnData: '0x5678' }
                      ]
                    ]
                  ),
                  logs: [],
                  gasUsed: '0x55b3',
                  status: '0x1'
                }
              ],
              difficulty: '0x0',
              excessBlobGas: '0x3220000',
              extraData: '0x',
              gasLimit: '0x1c9c380',
              gasUsed: '0x55b3',
              hash: '0x4342f0ab870175757176d5888e82f27523a6e1d52c0ac20a6ddf599d54ce0e04',
              logsBloom:
                '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
              miner: '0x95222290dd7278aa3ddd389cc1e1d165cc4bafe5',
              mixHash: '0x0000000000000000000000000000000000000000000000000000000000000000',
              nonce: '0x0000000000000000',
              number: '0x1455a09',
              parentBeaconBlockRoot: '0x0000000000000000000000000000000000000000000000000000000000000000',
              parentHash: '0xb30e288a5518544cc71dd24389a21061adab20f45f17c0907054dccf7bf00c01',
              receiptsRoot: '0x4ce78e593fcb88a20d7bb31c27879f800551f4069371e576f7efad0d9615a960',
              sha3Uncles: '0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347',
              size: '0x29b',
              stateRoot: '0xcfb988bd139a3f44aa79f9cbaba606429d76749fd18e1e8ca9a6612a4e0c8384',
              timestamp: '0x674f0627',
              totalDifficulty: '0xc70d815d562d3cfa955',
              transactions: params[0].blockStateCalls[0].calls.map(
                () => '0x96c6b7b5b4835fd518fd914feb586452cc797d332f1bd234f72c9e692c4c427a'
              ),
              transactionsRoot: '0xebe30ac336ac6714af65c684e278799f70965b18ae12e9b368056640ac111650',
              uncles: [],
              withdrawals: [],
              withdrawalsRoot: '0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421'
            }
          ]
        } else {
          // return a failure response
          return [
            {
              baseFeePerGas: '0x0',
              blobGasUsed: '0x0',
              calls: params[0].blockStateCalls[0].calls.map(() => ({
                returnData: '0x',
                error: {
                  message: 'foo',
                  code: 1,
                  data: viem.encodeErrorResult({
                    abi: IERC7412.abi,
                    errorName: 'OracleDataRequired',
                    args: ['0x2345234523452345234523452345234523452345', errorCode, BigInt(0)]
                  })
                },
                gasUsed: '0x55b3',
                status: '0x0'
              })),

              difficulty: '0x0',
              excessBlobGas: '0x3220000',
              extraData: '0x',
              gasLimit: '0x1c9c380',
              gasUsed: '0x55b3',
              hash: '0x4342f0ab870175757176d5888e82f27523a6e1d52c0ac20a6ddf599d54ce0e04',
              logsBloom:
                '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
              miner: '0x95222290dd7278aa3ddd389cc1e1d165cc4bafe5',
              mixHash: '0x0000000000000000000000000000000000000000000000000000000000000000',
              nonce: '0x0000000000000000',
              number: '0x1455a09',
              parentBeaconBlockRoot: '0x0000000000000000000000000000000000000000000000000000000000000000',
              parentHash: '0xb30e288a5518544cc71dd24389a21061adab20f45f17c0907054dccf7bf00c01',
              receiptsRoot: '0x4ce78e593fcb88a20d7bb31c27879f800551f4069371e576f7efad0d9615a960',
              sha3Uncles: '0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347',
              size: '0x29b',
              stateRoot: '0xcfb988bd139a3f44aa79f9cbaba606429d76749fd18e1e8ca9a6612a4e0c8384',
              timestamp: '0x674f0627',
              totalDifficulty: '0xc70d815d562d3cfa955',
              transactions: params[0].blockStateCalls[0].calls.map(
                () => '0x96c6b7b5b4835fd518fd914feb586452cc797d332f1bd234f72c9e692c4c427a'
              ),
              transactionsRoot: '0xebe30ac336ac6714af65c684e278799f70965b18ae12e9b368056640ac111650',
              uncles: [],
              withdrawals: [],
              withdrawalsRoot: '0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421'
            }
          ]
        }
      } else if (method === 'eth_call') {
        if (params[0].data === viem.encodeFunctionData({ abi: IERC7412.abi, functionName: 'oracleId' })) {
          // getOracleId call
          return viem.encodeFunctionResult({
            abi: IERC7412.abi,
            functionName: 'oracleId',
            result: viem.stringToHex('FAKE', { size: 32 })
          })
        } else if (params[0].data.includes('87651234') === true) {
          return viem.encodeAbiParameters(
            [viem.parseAbiParameter('(bool success, bytes returnData)[]')],
            [
              [
                { success: true, returnData: '0x1234' },
                { success: true, returnData: '0x5678' }
              ]
            ]
          )
        } else {
          /* eslint @typescript-eslint/no-throw-literal: "off" */
          throw {
            data: viem.encodeErrorResult({
              abi: IERC7412.abi,
              errorName: 'OracleDataRequired',
              args: ['0x2345234523452345234523452345234523452345', errorCode, BigInt(0)]
            })
          }
        }
      } else {
        return '0x'
      }
    })
  })

  describe('simulateWithOffchainData()', () => {
    it('passes a call execution error if its not recognized', async () => {
      const origError = new Error('0x08273020')
      fakeWeb3.request.mockRejectedValue(origError)
      await expect(
        async () =>
          await mod.simulateWithOffchainData(fakeWeb3, fakeAdapters, [
            { to: '0x1234123412341234123412341234123412341234', data: '0x12345678' }
          ])
      ).rejects.toThrowErrorMatchingSnapshot()
    })

    it('resolves offchain data and returns correct data', async () => {
      expect(
        await mod.simulateWithOffchainData(fakeWeb3, fakeAdapters, [
          { to: '0x1234123412341234123412341234123412341234', data: '0x12345678' },
          { to: '0x1234123412341234123412341234123412341234', data: '0x23456789' }
        ])
      ).toMatchObject([
        { returnData: '0x1234', success: true },
        { returnData: '0x5678', success: true }
      ])
    })

    it('fails if call repeat exceeded', async () => {
      // modify the error code so that the OracleDataRequired error continues forever.
      errorCode = '0x5757'

      await expect(
        mod.simulateWithOffchainData(fakeWeb3, fakeAdapters, [
          { to: '0x1234123412341234123412341234123412341234', data: '0x12345678' },
          { to: '0x1234123412341234123412341234123412341234', data: '0x23456789' }
        ])
      ).rejects.toThrow(new Error('erc7412 callback repeat exceeded'))
    })

    it('fails if no data in call response', async () => {
      fakeWeb3.request.mockResolvedValue(undefined)

      await expect(
        mod.simulateWithOffchainData(fakeWeb3, fakeAdapters, [
          { to: '0x1234123412341234123412341234123412341234', data: '0x12345678' },
          { to: '0x1234123412341234123412341234123412341234', data: '0x23456789' }
        ])
      ).rejects.toThrow(new Error('missing return data from multicall'))
    })
  })
})
