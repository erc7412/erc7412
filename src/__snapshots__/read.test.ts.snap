// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`read.ts simulateWithOffchainData() passes a call execution error if its not recognized 1`] = `
"An unknown RPC error occurred.

Details: 0x08273020
Version: viem@2.29.0"
`;
